### PR TITLE
EN-5103 - added first party mode

### PR DESCRIPTION
**Pull request by nitzanpx**

### CODE REVIEW

File: lib/px/pxconfig.lua [hunk 1]
```diff
@@ -55,6 +56,11 @@ _M.px_port = 443
 _M.ssl_enabled = true
 _M.enable_server_calls = true
 _M.send_page_requested_activity = true
+_M.base_url = _M.px_appId == 'PX_APP_ID' and 'sapi.perimeterx.net' or string.format('sapi-%s.perimeterx.net', _M.px_appId)
```
Comment (yaronpx) [inline lib/px/pxconfig.lua]: I don't like this. It's not really readable and will confuse customers trying to input their app id. also, it looks like you now ask the customer to input app id twice

File: lib/px/utils/pxclient.lua [hunk 2]
```diff
@@ -160,6 +165,74 @@ function M.load(config_file)
...
+            end
+        end
+
+        httpc:set_timeout(2000)
+        httpc:proxy_response(httpc:proxy_request())
+        httpc:set_keepalive()
+    end
+
+    function _M.reverse_px_client()
+        local px_request_uri = "/" .. px_config.px_appId .. "/main.min.js"
```
Comment (yaronpx) [inline lib/px/utils/pxclient.lua]: shouldn't this be the appId without the 'PX' prefix?

Comment (nitzanpx) [reply lib/px/utils/pxclient.lua]: no, this is where to redirect the requset to

File: lib/px/pxnginx.lua [hunk 3]
```diff
@@ -90,6 +90,21 @@ function M.application(file_name)
             return true
         end
     end
+
+    if px_config.first_party_enabled then
+        local reverse_prefix = string.sub(px_config.px_appId, 3, string.len(px_config.px_appId))
+        local lower_request_url = string.lower(ngx.var.request_uri)
+        if string.find(lower_request_url, string.lower("/" .. reverse_prefix .. px_constants.FIRST_PARTY_VENDOR_PATH)) then
```
Comment (yaronpx) [inline lib/px/pxnginx.lua]: add some comments here, as it's not trivial

File: lib/px/utils/pxconstants.lua [hunk 4]
```diff
@@ -5,12 +5,16 @@
...
+    MODULE_VERSION = "NGINX Module v3.1.0",
     RISK_PATH = "/api/v2/risk",
     CAPTCHA_PATH = "/api/v2/risk/captcha",
     ACTIVITIES_PATH = "/api/v1/collector/s2s",
     TELEMETRY_PATH = "/api/v2/risk/telemetry",
-    REMOTE_CONFIGURATIONS_PATH = "/api/v1/enforcer"
+    REMOTE_CONFIGURATIONS_PATH = "/api/v1/enforcer",
+    ENFORCER_TRUE_IP_HEADER = 'x-px-enforcer-true-ip',
+    FIRST_PARTY_HEADER = 'x-px-first-party',
+    FIRST_PARTY_VENDOR_PATH = '/main.min.js',
```
Comment (yaronpx) [inline lib/px/utils/pxconstants.lua]: I think this changed in the design

File: lib/px/pxconfig.lua [hunk 5]
```diff
@@ -55,6 +58,11 @@ _M.px_port = 443
...
+_M.collector_host = _M.px_appId == 'PX_APP_ID' and 'collector.perimeterx.net' or string.format('collector-%s.perimeterx.net', _M.px_appId)
+_M.client_host = "client.perimeterx.net"
+_M.collector_port_overide = nil
```
Comment (dimapx) [inline lib/px/pxconfig.lua]: typo - `override`
on `client_port_overide` as well

File: lib/px/utils/pxconstants.lua (hunk 4 above)
Comment (nitzanpx) [reply lib/px/utils/pxconstants.lua]: not yet

**Pull request merged**